### DISCUSSION TITLE
Question about untrusted user input

### DISCUSSION THREAD

**Discussion by madjar** [oxigraph/oxigraph#253] [Q&A] [answered]
Hi there!

I looking at the documentation for oxigraph, as I am considering using it for a small project, and I wondered if you might have some thoughts on this: how to handle (potentially untrusted) user input in SPARQL queries?

The two approaches I usually see in SQL are

- To use some form of placeholders in the query (eg WHERE id = ?). It doesn't seem to be an option in oxigraph.
- String concatenation with explicit escaping (though the previous option is often preferred). A few hits on stackoverflow (https://stackoverflow.com/questions/29601839/standard-regex-to-prevent-sparql-injection) suggest that this can be done in sparql by using triple quotes and escaping all backslashes, and single and double quotes.

Is that a topic you've encountered? If so, how do you usually deal with this?

**Comment by Tpt** [comment, marked as answer]
Hi! Thank you for trying oxigraph.

If you just want to allow unterested values inside of SPARQL queries, escaping is indeed the best solution. Oxigraph does not support (yet?) placeholders. To do validation and escaping the easiest way is to use Oxigraph model objects (`NamedNode`, `Literal`...) imho. These objects validates the values when created and their "to_string" mechanisms properly escapes the values.

In Rust:
```rust
NamedNode::new(MY_INPUT)?.to_string()
```

In Python (will raise an exception if the input is not a valid IRI):
```rust
str(NamedNode(MY_INPUT))
```

`BlankNode` and `Literal` works similarly.

**Comment by madjar** [comment]
Thanks a lot, it is very helpful to know that oxigraph provides some way to escape, so I don't have to roll my own escaping code!
